### PR TITLE
[FLINK-9304] Timer service shutdown should not stop if interrupted

### PR DESCRIPTION
**Pull request by StefanRRichter**
## What is the purpose of the change

This PR prevents that interruption prematurely stops the shutdown of the timer service.

## Brief change log

We catch `InterruptedException` around the shutdown call and repeat the attempt while there is still time left w.r.t. to the shutdown grace period.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @tillrohrmann

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 1]
```diff
@@ -706,6 +689,38 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
...
+
+				final long totalTimeoutMs = getEnvironment().getTaskManagerInfo().getConfiguration().
+					getLong(TaskManagerOptions.TASK_CANCELLATION_TIMEOUT_TIMERS);
+
+				final Deadline deadline = Deadline.fromNow(Duration.ofMillis(totalTimeoutMs));
+
+				boolean timerServiceShutdownComplete = false;
+
+				while (!timerServiceShutdownComplete && deadline.hasTimeLeft()) {
+					try {
```
Comment (sihuazhou) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java]: Shall we at least try to shutdown the time service one time? If so, I think maybe we should use `do {...} while()` here.

**Comment by StephanEwen** [comment]
The comment to make at least one attempt sounds good.

If you refactor this a little bit, you can actually test it. Either make it a static method to which you pass the timer service, or better, make it a method `shutdownUninterruptible(Deadline, Logger)` on the TimerService itself.

**Comment by StefanRRichter** [comment]
Thanks for the comments, I have updated the PR with another commit. Please take a look again if this can be merged.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeService.java [hunk 2]
```diff
@@ -197,6 +204,23 @@ public boolean shutdownAndAwaitPending(long time, TimeUnit timeUnit) throws Inte
 		return timerService.awaitTermination(time, timeUnit);
 	}
 
+	@Override
+	public boolean shutdownServiceUninterruptible(long timeoutMs) {
```
Comment (tillrohrmann) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeService.java]: Do we want to make the shutdown of the timer service swallow the `InterruptedException` also for all subsequent calls happening after the shutdown of the timer service?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeServiceTest.java [hunk 3]
```diff
@@ -504,7 +475,101 @@ public void testShutdownAndWaitPending() {
...
+				try {
+					Thread.sleep(10);
+				} catch (InterruptedException ignore) {
+				}
+			}
+		});
+
+		interruptCallerThread.start();
+
+		final long timeoutMs = 1000L;
```
Comment (tillrohrmann) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeServiceTest.java]: Can we make this a bit faster?

Comment (StefanRRichter) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeServiceTest.java]: 👍

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeService.java (hunk 2 above)
Comment (StefanRRichter) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeService.java]: 👍

**Pull request closed without merging**